pkg/controller/types/group.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2019 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"github.com/vmware-tanzu/antrea/pkg/apis/controlplane"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type GroupType string
+
+const (
+	ClusterGroup    GroupType = "ClusterGroup"
+	NamespacedGroup GroupType = "NamespacedGroup"
+)
+
+type GroupReference struct {
+	// Type of the Group.
+	Type GroupType
+	// Namespace of the Group. It's empty for ClusterGroup.
+	Namespace string
+	// Name of the Group.
+	Name string
+	// UID of the Group.
+	UID types.UID
+}
+
+// GroupSelector describes how to select GroupMembers.
+type GroupSelector struct {
+	// The normalized name is calculated from Namespace, PodSelector, ExternalEntitySelector and NamespaceSelector.
+	// If multiple policies have same selectors, they should share this group by comparing NormalizedName.
+	// It's also used to generate Name and UUID of group.
+	NormalizedName string
```

Comment:
Do you think it helps a lot to share one group if selectors are the same, or maybe just keep each group separate (but can still optimize for standalone selectors not in any group)?
@tnqn

Reply:
i have not changed the way we handle standalone selectors and did not convert them in to Groups first. 
for Groups, we could switch to not sharing them for the purpose of debuggability, in which case I will use the K8s generated UIDs of Groups for internal types. lets hear from @tnqn too.

Reply:
I think if the we don't share internal groups between ClusterGroup and namespaced Group resources in the future (which might be very hard to implement, i.e. figuring out if a CG's namespaceSelector <=> Group's namespace), there might not be a lot of advantage to this approach. Standalone selectors might come repeated a lot in multiple networkpolicies, but I guess it would not be common where a few ClusterGroups have the exact same selectors in their definitions. @tnqn

Reply:
@Dyanngg : my original comment is not about sharing internal group struct - I have no issue with that; but should we optimize and let two groups with the same selectors to share one group instance.

Reply:
> @Dyanngg : my original comment is not about sharing internal group struct - I have no issue with that; but should we optimize and let two groups with the same selectors to share one group instance.

I assumed this "one group instance" you mentioned here was the "internal group". We maybe can't/shouldn't stop user from creating ClusterGroup CRD objects which has the same selectors - my point is that seems to be a quite unusual case.

Reply:
Yes, we cannot disallow users from creating groups with the same selectors. But in our implementation, we can choose to always create two separate internal groups for two groups even they share the same selectors?

Reply:
> Yes, we cannot disallow users from creating groups with the same selectors. But in our implementation, we can choose to always create two separate internal groups for two groups even they share the same selectors?

Yes it is what the latest implementation, and I am 100% on board with not sharing internal groups. Sorry I might have confused you by the comment earlier.

Reply:
Great we are on the same page. @tnqn : do you have any opinion on this?

Reply:
The current implementation in the PR is as follows:
1. For ClusterGroup with same selectors, do not share internal groups
2. For standalone selectors in policies, share internal groups
Seems everyone is on the same page.

Reply:
I feel keeping 1:1 mapping is easier to understand and debug. Does it simplify the implementation and reduce computation too?

Reply:
actually keeping 1:1 makes the implementation simpler as we no longer need to check whether internal groups are being referenced before deleting and generation of UIDs based on selector.

Reply:
But I do feel later we might consider some optimization for Namespaces, as many groups and selectors can share one Namespace.